### ISSUE TITLE
Controlling the size of data batches sent to the custom loss function

### ISSUE THREAD

**Issue by omer-shaked** [catboost/catboost#624]
catboost version: 0.10.3
Operating System: Ubuntu
CPU

Hi,

I'm trying to use a custom loss function that aggregates across multiple samples (the derivatives of one sample are affected by the derivatives of other sample/s).
According to the answer given to issue #521 (question 3), I see that the intention is indeed for the function to receive the approxes values of the entire data set in a single batch.

However, when I perform training I see that data is being sent to the custom loss function in mini-batches (example is given below).

So, I would be happy to know if this is the expected behavior, and if there's a way to modify this behavior in order to guarantee that all samples arrive together in a single call for the loss function.

Thanks !
Omer

Example:
In a data set with 3,551 samples, I print from within the function the number of samples that arrive in every function call (all from the same training iteration):

INFO - 2019-01-17 13:45:25,483 custom_train_functions:40 calc_ders_range Num of samples: 284
INFO - 2019-01-17 13:45:25,666 custom_train_functions:40 calc_ders_range Num of samples: 1000
INFO - 2019-01-17 13:45:25,746 custom_train_functions:40 calc_ders_range Num of samples: 568
INFO - 2019-01-17 13:45:25,859 custom_train_functions:40 calc_ders_range Num of samples: 142
INFO - 2019-01-17 13:52:06,631 custom_train_functions:40 calc_ders_range Num of samples: 136
INFO - 2019-01-17 13:52:06,940 custom_train_functions:40 calc_ders_range Num of samples: 1000
INFO - 2019-01-17 13:52:07,096 custom_train_functions:40 calc_ders_range Num of samples: 1000
INFO - 2019-01-17 13:52:16,061 custom_train_functions:40 calc_ders_range Num of samples: 272
INFO - 2019-01-17 13:52:16,142 custom_train_functions:40 calc_ders_range Num of samples: 1000
INFO - 2019-01-17 13:52:16,248 custom_train_functions:40 calc_ders_range Num of samples: 1000
INFO - 2019-01-17 13:52:20,755 custom_train_functions:40 calc_ders_range Num of samples: 551
INFO - 2019-01-17 13:52:20,801 custom_train_functions:40 calc_ders_range Num of samples: 1000
INFO - 2019-01-17 13:52:21,171 custom_train_functions:40 calc_ders_range Num of samples: 2
INFO - 2019-01-17 13:52:21,219 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,255 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,259 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,287 custom_train_functions:40 calc_ders_range Num of samples: 2
INFO - 2019-01-17 13:52:21,313 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,341 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,392 custom_train_functions:40 calc_ders_range Num of samples: 2
INFO - 2019-01-17 13:52:21,415 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,441 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,509 custom_train_functions:40 calc_ders_range Num of samples: 2
INFO - 2019-01-17 13:52:21,532 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,553 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,582 custom_train_functions:40 calc_ders_range Num of samples: 1
INFO - 2019-01-17 13:52:21,653 custom_train_functions:40 calc_ders_range Num of samples: 1
and so on ...

**Comment by david-waterworth** [comment]
Is it not sending the entire batch for the current tree node? although the 1000 looks suspiciously like a hard limit!

**Comment by Noxoomo** [comment]
Hi!

May be this will solve issue: 

change boosting_type option to Plain and set option has_time to True. 
Otherwise CatBoost could reorder inputs and will use ordered-scheme (see our NIPS paper for details https://papers.nips.cc/paper/7898-catboost-unbiased-boosting-with-categorical-features.pdf), for which we'll need to compute gradients of prefixes of dataset

**Comment by omer-shaked** [comment]
@david-waterworth  and @Noxoomo  thanks for your replies.

I ran the custom loss function again, this time on a set with 38,560 samples, and set the boosting_type and has_time as suggested by @Noxoomo .

I still received the data is small batches of no more than 1000 samples at a time (logs are attached below).
However, the data arrived in order, which at least in my case puts me in a better situation (because I only need to know the values of 'nearby' samples for the calculation that I perform).

I'll be happy to know if you have any other suggestions on how to still try and force all the data to arrive in a single batch.

In addition, I was surprised to see that in each iteration, first all the data arrives in batches of 1000 samples (and a last batch of 560 samples), but then additional batches of ~300 samples arrive at the same iteration.
I assumed that in 'Plain' mode the derivatives are calculated once per sample per iteration, but this appears not to be the case. Can you explain to me what am I missing here ?
[custom_loss.log](https://github.com/catboost/catboost/files/2776903/custom_loss.log)

**Comment by annaveronika** [comment]
Yes, if you change boosting_type option to Plain and set option has_time to True, you will still get data in blocks - this is because these blocks are calculated in parallel on different threads.
If you want to make it in one portion you need to set block size to size of your dataset. If you still want that, we can look into code to find this place -  let us know.

As for batches of size 1000 and of size 300 - we have to check, I don't have an answer right now.

**Comment by omer-shaked** [comment]
Hi @annaveronika ,

I've managed to get around this to some degree by using the fact that with 'has_time' = True the samples are ordered, and so I do the loss calculation as if all samples relevant for each calculation are found in the same batch of 1000 samples.
However, since this is an approximation that is not accurate, and since I believe having all samples in a single batch can reduce the run times - I'll be happy if this issue is fixed (assuming this is not a complicated fix).

**Comment by annaveronika** [comment]
@omer-shaked This is not a bug, this is what we actually want - we want different blocks to be processed simultaneously by different threads. We don't plan to change this behavoir. We can only point you to a place in code where you can change BlockSize to the size of the train dataset for you to have a hack for your code to work.
@Evgueni-Petrov-aka-espetrov Evgueni, could you please help us with that?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Hi @omer-shaked,

You may set the desired block size here
https://github.com/catboost/catboost/blob/453dd71429f0ddf055eb61c9b79b1f564ce94ffa/catboost/libs/algo/approx_calcer.cpp#L101

A block size larger than the number of samples will disable threading.

Evgueni

**Comment by annaveronika** [comment]
@omer-shaked Let us know if this helps

**Comment by omer-shaked** [comment]
Thanks @annaveronika @Evgueni-Petrov-aka-espetrov .
This is in principle exactly what I need.
However, since I use the python package - is there a way to expose this parameter as part of the python API ?

**Comment by annaveronika** [comment]
These are details of speed tuning, which are very much internal, no one should care about them.
You needed this one to make a hack to calculate derivatives using all the data. We don't support this behavoir, for custom losses we only support pointwise errors that require knowledge about a single object or about a group of objects if groups are present. Thus we don't want to expose this particular implementation detail.

**Comment by omer-shaked** [comment]
Ok, thanks for the explanation. I'll modify the cpp code locally if I'll find it extremely neccessary.

On your side, I would consider adding this limitation to the documentation of the custom loss functions ("for custom losses we only support pointwise errors ...")

Besides that, as far as I'm concerned you are welcome to close this item.

**Comment by annaveronika** [comment]
We don't plan to do anything here, so I think I should close the issue.

**Comment by jhogg11** [comment]
> catboost/catboost/libs/algo/approx_calcer.cpp

Is it possible to do this in Python rather than altering the cpp file?

Sort of an unrelated question that might provide an alternative way to deal with all samples at once, but is it possible to alter the internal parameters of a trained model after it's been trained (e.g., through a stochastic optimization algorithm)?  If so, which values need to be altered?

For example, I know it's possible to train a neural network with something like particle swarm optimization and I'm wondering if something similar can be done with CatBoost.  I'm just not sure exactly which attributes actually define the model in terms of the predictions that it makes, assuming the structure has already been determined.

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
hi @jhogg11 🤝 
thank you very much for trying catboost 🔥 

`APPROX_BLOCK_SIZE` is introduced to improve L1-cache utilization, there is no sense to tune it
the following example gives an idea of how to train models using batches
https://catboost.ai/en/docs/concepts/python-usages-examples#batch-training

catboost/lgbm/xgb models are sums of decision trees
in a decision tree of a fixed structure, we may tune split boundaries and leaf values
tuning these parameters in a catboost model would be similar to training a random forest
we did not research whether this way we get some significant advantages in prediction accuracy
